### ISSUE TITLE
Question: Signaling over WebSocket

### ISSUE THREAD

**Issue by nicolastmaia** [sipsorcery-org/sipsorcery#703]
Hey there.
First of all, thanks for creating SipSorcery, it's a really nice and powerful tool.

So, I'm new to this whole media communication thing, but was hoping to get a C# application to send audio to a web browser application with both of them behind NAT. 

The flow of my application is: client (in browser) hits "Start" and server (C# application) starts sending the audio to it. I am already able to send audio from a buffer via stream and everything works in LAN, but can't seem to get it to work when behind NAT.
I've been checking and adapting the examples here, but it seems I can't get signaling to work. No matter what I do, the C# side always gets stuck in "have_local_offer" or "stable" state.
The versions I've been "messing around with" are not commited, but I could get the STUN and TURN configurations to work. I really think signaling is my problem.

Here is the [repo](https://github.com/nicolastmaia/WebRTCLearn) I've been using to practice. The master branch is the version that works over LAN but not if behind NAT.
It would be great if you could take a look and give me some pointer.

Thanks a lot.

**Comment by sipsorcery** [comment]
In you `client.html` you have `const WEBSOCKET_URL = "ws://127.0.0.1:8081/"`. That's only ever going to work if the web socket server is on the same machine as the client. It will definitely not work across a network let alone behind a NAT.

The first step you need to do is make sure your web socket server is accessible from the client machine. To do that you can use something like telnet to check if the port is rechable, e.g:

`telnet 1.1.1.1 8081` (and replace 1.1.1.1 with the IP address of your web socket server machine)

If the web socket server machine is behind a NAT/firewall you'll potentially need to create a port forwarding and/or firewall rule to make it reachable. If you have an IPv6 address that is often easier than dealing with NAT.

Hopefully this is all making sense. If not, you've got a bit of reading to do :).

**Comment by nicolastmaia** [comment]
Thanks for the response.

I only used that WEBSOCKET_URL constant when I was testing it locally. When on the internet I have a domain name for the server that I use to connect client and server. 

The telnet command is working and so is my client connection to the websocket when behind NAT (the server logs events that show they are communicating). The problem is that this communication doesn't get past the "have_local_offer" or "stable" step on the server.

Summarizing, the problem is I can't figure out what to do for the ICE candidates and offer exchange part. Does RTCPeerConnection do those under the hood or do I have to explicitly do it?

**Comment by sipsorcery** [comment]
The RTCPeerConnection will gather all the local candidates and send them in its initial offer or answer. If you're relying on STUN or TURN ICE candidates you may need to take some extra steps.

It's best to rely on the browser sending its STUN or TRUN candidates to your app. The webrtc stack in teh browser is a lot more sophisticated than this library. If that's not an option you can try the snippet below to forward candidates from your app to the browser once they are discovered, which will only be the STUN or TURN ones:

````
pc.onicecandidate += (candidate) =>
{
    if (pc.signalingState == RTCSignalingState.have_local_offer ||
        pc.signalingState == RTCSignalingState.have_remote_offer)
        {
            context.WebSocket.Send($"candidate:{candidate}");
        }
};
````

**Comment by nicolastmaia** [comment]
> It's best to rely on the browser sending its STUN or TRUN candidates to your app

Shouldn't I gather STUN and TURN candidates on both the browser and sipsorcery app?

Something I noticed recently is that the ice candidates that both app and browser send to each other don't include their public IPs. I checked that the STUN server returns candidates with the public IPs of each side respectively, but they don't get sent to the other party. This is my current suspicion on what might be wrong.

All candidates are like this:
![image](https://user-images.githubusercontent.com/45211638/159910654-27e2dd67-cc01-4bf5-9564-bd72ba3b2e49.png)

**Comment by nicolastmaia** [comment]
Hey guys.
Just got all pieces together and audio transmission is fully functional!

I'd just like to say thank you for trying to help me and to register here what was wrong so it might help others.

Basically, the code on both sides of the communication stayed the same as if not behind NAT (WebRTCGetStartedExample). No need to write any code to send ice candidates or offer/answer on the sipsorcery side, the lib does it all internally (at least for my use case). On the client side I used the code available in the example WebRTCGetStarted here in the repo. The only thing I had to add on both sides was the code to pass the STUN/TURN server when creating an RTCPeerConnection:

```      
const config = {
  iceServers: [{
    urls: 'STUN_TURN_URL', // stun example: 'stun:your_stun_server_here:your_stun_port_here'; turn example: 'turn:your_turn_server_here:your_turn_port_here'
    username: 'SERVER_USERNAME',
    credentialType: "password",
    credential: 'SERVER_PASSWORD'
  }]
};

pc = new RTCPeerConnection(config);
```

The problem I was facing had to do with the stun/turn servers used. On the client side I could communicate with the stun/turn and get my public IP, but there was some block on the VM I put sipsorcery in and it never got its public IP from the STUN server. For that reason, webrtc wasn't matching candidates and so I couldn't connect. Fixed the stun connection problem and now the sipsorcery app can communicate with the stun server and get its public IP (also tried with a google public STUN and it worked). Everything working.

One last thing I noticed is that for it to work I *had* to put the TURN address on the client and for the server only the stun was enough. I understand that this is because, in my case, the server is behind symmetric NAT and the client is behind permissive NAT, but still don't get why ONLY the client needed to use the TURN address. Got a bit of reading to do, but at least now they are talking.

Some hints I picked up along the way:
- This is the [link](https://clients.dh2i.com/NatTest/) I used to find out my kind of NAT on both sides; 
- This is the [link](https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/) I used to test my STUN/TURN servers. 
- To check the ice candidate matches just open firefox and go to about:webrtc, all webrtc communication on other tabs (of firefox) will get recorded there and you'll see ice candidates exchanged and matched.
- This [link](https://www.cloudbees.com/blog/webrtc-issues-and-how-to-debug-them) has great hints on how to debug WebRTC.

Thanks again for all the help and attention!
